### PR TITLE
Extract instrumentation logic out of DefaultCachedClasspathTransformer

### PR DESCRIPTION
**Pull request by asodja**
So we can reuse all instrumentation logic (excluding caching and wiring) in the artifact transform. There were no changes to actual logic.

And expose `JarTransformFactoryForAgent` and `JarTransformFactoryForLegacy` as a service, so we can inject them in artifact transform.

All extracted logic now lives in `org.gradle.internal.classpath.transforms` package. 

I also did some renaming, so class names are more specific: 
- `InstrumentingClasspathFileTransformer.Transformation` is now called `JarTransform`
- `CachedClasspathTransformer.Transform` is now `ClassTransform`
- `InstrumentingClasspathFileTransformer.Policy` is now `JarTransformFactory` since it creates `JarTransform`

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by asodja** [comment]
@bot-gradle test this

**Review by h0tk3y — APPROVED** [review]
LGTM, but I would suggest waiting for @mlopatkin's review, too, as he probably has a better view of this piece of infrastructure.

**Review by h0tk3y — APPROVED** [review]
LGTM, but I would suggest waiting for @mlopatkin's review, too, as he probably has a better view of this piece of infrastructure.

**Comment by lptr** [comment]
Would it make sense to move instrumentation-related code out of `:core`? Could we move it to one of the `:instrumentation-*` subprojects? Even creating a new one seems like an improvement to me over leaving it in the big bag of classes that's `:core`.

File: subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/JarTransformFactory.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import org.gradle.internal.hash.Hasher;
+
+import java.io.File;
+
+/**
+ * Jar transform factory. There are some differences when instrumenting classes to be loaded by the instrumenting agent, this interface encapsulates them.
+ */
+public interface JarTransformFactory {
+    /**
+     * Modifies JAR content hash according to the algorithm implemented by this policy.
```
Comment (mlopatkin) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/JarTransformFactory.java:29]: :nail_care: this still uses the old `policy` name, maybe reword it like this:
```suggestion
     * Updates jar/directory content hash to account for the transform produced by this factory.
```

File: subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/JarTransformForAgent.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+import java.io.IOException;
+import java.util.jar.Attributes;
+import java.util.jar.Manifest;
+
+import static org.gradle.internal.classpath.transforms.MrJarUtils.isSupportedVersion;
+
+/**
+ * Transformation for agent-based instrumentation.
+ */
+public class JarTransformForAgent extends BaseJarTransform {
```
Comment (mlopatkin) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/JarTransformForAgent.java:40]: :nail_care: can the `JarTransformForAgent` class be package-private?

File: subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/JarTransformFactoryForAgent.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    @Override
+    public JarTransform createTransformer(File file, ClassTransform classTransform, InstrumentingTypeRegistry typeRegistry) {
+        return new JarTransformForAgent(file, classpathBuilder, classpathWalker, typeRegistry, classTransform);
+    }
+
+    @Override
+    public String toString() {
+        return "Policy(agent)";
```
Comment (mlopatkin) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/JarTransformFactoryForAgent.java:53]: :nail_care: `Policy` -> `TransformFactory`.

File: subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/JarTransformFactoryForLegacy.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        return new BaseJarTransform(source, classpathBuilder, classpathWalker, typeRegistry, classTransform);
+    }
+
+    private boolean isJarSignatureFile(String entryName) {
+        return entryName.startsWith("META-INF/") && entryName.endsWith(".SF");
+    }
+
+    @Override
+    public String toString() {
+        return "Policy(legacy)";
```
Comment (mlopatkin) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/JarTransformFactoryForLegacy.java:92]: 💅 `Policy` -> `TransformFactory`.

File: subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/JarTransformFactoryForLegacy.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    /**
+     * A no-op transformation that copies the original file verbatim. Can be used if the original cannot be instrumented under policy.
+     */
+    private static class SkipJarTransform implements JarTransform {
+
+        private static final Logger LOGGER = LoggerFactory.getLogger(InstrumentingClasspathFileTransformer.class);
```
Comment (mlopatkin) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/JarTransformFactoryForLegacy.java:100]: :nail_care: having a dependency on `InstrumentingClasspathFileTransformer` just for the logger seems excessive. Can we use something else?

File: subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/MultiReleaseJarTransformForLegacy.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+import static org.gradle.internal.classpath.transforms.MrJarUtils.isInUnsupportedMrJarVersionedDirectory;
+
+/**
+ * Transformation for legacy instrumentation when transformed JARs are part of the classpath.
+ * This is still used when TestKit and TAPI run Gradle in embedded or debug mode.
+ * <p>
+ * This transformation filters out not yet supported versioned directories of the multi-release JARs.
+ */
+public class MultiReleaseJarTransformForLegacy extends BaseJarTransform {
```
Comment (mlopatkin) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/MultiReleaseJarTransformForLegacy.java:35]: :nail_care: can `MultiReleaseJarTransformForLegacy` be a package-private class?

File: subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/MrJarUtils.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    public static boolean isInUnsupportedMrJarVersionedDirectory(ClasspathEntryVisitor.Entry entry) {
+        OptionalInt version = JarUtil.getVersionedDirectoryMajorVersion(entry.getName());
+        if (version.isPresent()) {
+            return !isSupportedVersion(version.getAsInt());
+        }
+        // The entry is not in the versioned directory at all.
+        return false;
+    }
+
+    public static boolean isSupportedVersion(int javaMajorVersion) {
```
Comment (mlopatkin) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/MrJarUtils.java:43]: :nail_care: I'm not sure if this is a good place for this method, it isn't related to JARs at all, maybe move it to `AsmConstants` instead, as it matches its semantics (is the given java version supported by ASM)?

**Review by mlopatkin — APPROVED** [review]
Looks good except for a few cosmetic and naming suggestions.

:thinking: I don't like that we name this things `JarXxxx`, its input isn't necessary a JAR, but can be a directory with class files. Maybe `ClasspathElement` is a better name, and we use `element` for this purpose  in some places already. `ClasspathEntry` is also an option, but we already use `Entry` for individual class files and there's potential for a confusion with `JarEntry`, so I'd rather avoid this term.

File: subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/JarTransformForAgent.java (hunk 2 above)
Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/JarTransformForAgent.java:40]: It can be, fixed

File: subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/JarTransformFactoryForAgent.java (hunk 3 above)
Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/JarTransformFactoryForAgent.java:53]: Fixed

File: subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/JarTransformFactoryForLegacy.java (hunk 4 above)
Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/JarTransformFactoryForLegacy.java:92]: Fixed

File: subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/JarTransformFactoryForLegacy.java (hunk 5 above)
Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/JarTransformFactoryForLegacy.java:100]: Good catch, we can just use SkipJarTransform

File: subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/MultiReleaseJarTransformForLegacy.java (hunk 6 above)
Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/MultiReleaseJarTransformForLegacy.java:35]: It can be, fixed

File: subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/MrJarUtils.java (hunk 7 above)
Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/MrJarUtils.java:43]: Moved to AsmConstants

File: subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/LambdaSerializationTransformer.java [hunk 8]
```diff
@@ -63,7 +63,7 @@ class LambdaSerializationTransformer extends ClassVisitor {
     private static final Type SERIALIZED_LAMBDA_TYPE = getType(SerializedLambda.class);
 
     private static final String RETURN_OBJECT_FROM_SERIALIZED_LAMBDA = getMethodDescriptor(OBJECT_TYPE, SERIALIZED_LAMBDA_TYPE);
-    private static final String RETURN_STRING = getMethodDescriptor(CommonTypes.STRING_TYPE);
+    private static final String RETURN_STRING = Type.getMethodDescriptor(CommonTypes.STRING_TYPE);
```
[the remark this answers is not in the document]
Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/LambdaSerializationTransformer.java]: Good question, removed

File: subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/BaseJarTransform.java [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.ClassWriter;
+
+import java.io.File;
+import java.io.IOException;
+
+/**
+ * Base class for the transformations. Note that the order in which entries are visited is not defined.
+ */
+public class BaseJarTransform implements JarTransform {
```
[the remark this answers is not in the document]
Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/BaseJarTransform.java]: It can be, fixed

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test and merge
